enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,545 @@
+
+When reconciling a PO custom resource, the PO manager will need to propagate the underlying BundleDeployment's status back to the PO custom resource that's currently being processed, and relaying that individual PO status information back to the "aggregate" ClusterOperator that controller is managing.
+
+### Workflow Description
+
+```mermaid
+sequenceDiagram
+actor Admin
+participant Installer
+participant Bootstrap
+participant CVO
+participant ClusterOperator
+participant PO manager
+Installer->>Admin: Creates the Kubernetes manifests
+CVO->>PO custom resource: Generates
+Admin-->>Installer: Edits Kubernetes manifests
+Installer->>Bootstrap: Creates
+Bootstrap->>CVO: Deploys
+CVO->>PO manager: Deploys
+PO manager->>ClusterOperator: Manages
+PO manager->>PO custom resource: Watches
+CVO->>ClusterOperator: Watches
+```
+
+Note: The above workflow requires admins to manually create, and edit, the cluster Kubernetes manifests with their desired PO intent. There's a chance that admins mistype the package names, and depending on whether upfront validation is possible, that invalid configuration may not be surfaced until after cluster creation. This behavior may require explicit documentation to reduce this risk. These are obvious UX gaps with this workflow, which can be solidified in future phases, but it will result in zero installer changes during the 4.12 timeframe.
+
+### API Extensions
+
+When the "TechPreviewNoUpgrades" feature set is enabled for a cluster, several API extensions will be added during phase 0.
+
+#### OpenShift API(s)
+
+> Note: The following API(s) will live in the openshift/api repository.
+
+- Introduces a top-level PlatformOperator custom-scoped v1alpha1 CRD:
+
+```go
+const (
+    TypeSourced = "Sourced"
+    TypeApplied = "Applied"
+
+    ReasonSourceFailed     = "SourceFailed"
+    ReasonSourceSuccessful = "SourceSuccessful"
+    ReasonApplyFailed      = "ApplyFailed"
+    ReasonApplySuccessful  = "ApplySuccessful"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:scope=Cluster
+
+// PlatformOperator is the Schema for the platformoperators API.
+type PlatformOperator struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   PlatformOperatorSpec   `json:"spec,omitempty"`
+    Status PlatformOperatorStatus `json:"status,omitempty"`
+}
+
+// PlatformOperatorSpec defines the desired state of PlatformOperator.
+type PlatformOperatorSpec struct {}
+
+// PlatformOperatorStatus defines the observed state of PlatformOperator.
+type PlatformOperatorStatus struct {
```

Comment:
This is very basic right now, do you intend to proxy any information about the deployed Bundles onto this status or are you expecting users to look that up? In which case, should there be a reference to anything this creates so that a user can see what was created as a result of this object existing?

Reply:
I think we'll have a standalone controller that's responsible for proxying state back to the CVO, and so we may want/need a dedicated "status.Ref" (naming TBD) field that references the underlying BundleDeployment's resource name that gets generated by the PO manager component during it's reconciliation loop.

Alternatively: we just bubble up the generated BundleDeployment resource's name in a status condition message w/o needing an auxiliary status.Ref field. The standalone controller could then just fire off LIST queries for the BundleDeployments in the cluster w/o needing to interface with any of the PO resources.

For the most part, having just status conditions seems like it could be sufficient for phase 0 despite the small API surface for this top-level PO API.

Reply:
A ref to the underlying bundledeployment sounds like it would be useful to a user, that's much easier to get out (using eg jq) than substring matching on a condition of a particular type

Reply:
@joelanford and I were talking about another ref field for that bundle name that was "sourced" (e.g. selected) from an admin configured package in order to avoid performing unnecessary work during phase 0 when we've already installed a PO.

What's typically our guidance on status ref keys? Is something like a corev1.ObjectReference preferable over a string value?

Reply:
You should create a struct of your own to reference the resource. We don't like `corev1.ObjectReference` as typically it has superfluous fields that don't make sense in our own APIs. You should create a struct with exactly the fields you want to expose and use that instead

Reply:
Added a status.activeBundleDeployment field in https://github.com/openshift/enhancements/pull/1170/commits/41663f9fbd414cc5ac1f6d2c600e06fd0c4a24db and updated the implementation details for how this field will be set.

Reply:
> were talking about another ref field for that bundle name that was "sourced" (e.g. selected) from an admin configured package in order to avoid performing unnecessary work during phase 0 when we've already installed a PO.

Ended up going into another direction from this implementation too. Introducing a status field for the sourced bundle name during phase 0, just to remove that implementation from the API in subsequent phases doesn't seem worth the churn. Instead, we can simply check whether the sourced status condition type has been set to metav1.ConditionTrue in the reconciliation logic.